src/JustSaying/Fluent/PublishConfig/DynamicMessagePublisher.cs

```diff
@@ -62,6 +65,48 @@ public async Task PublishAsync(Message message, PublishMetadata metadata, Cancel
         await config.Publisher.PublishAsync(message, metadata, cancellationToken).ConfigureAwait(false);
     }
 
+    /// <inheritdoc/>
     public Task PublishAsync(Message message, CancellationToken cancellationToken)
         => PublishAsync(message, null, cancellationToken);
+
+    /// <inheritdoc/>
+    public async Task PublishAsync(IEnumerable<Message> messages, PublishBatchMetadata metadata, CancellationToken cancellationToken)
+    {
+        var publisherTask = new List<Task>();
+        foreach (var group in messages.GroupBy(x => x.GetType()))
+        {
+            var batchMessages = group.ToList();
+            var message = batchMessages[0];
+
+            string topicName = _topicNameCustomizer(message);
+            if (_batchPublisherCache.TryGetValue(topicName, out var publisher))
+            {
+                publisherTask.Add(publisher.PublishAsync(batchMessages, metadata, cancellationToken));
+                continue;
+            }
+
+            var lockObj = _topicCreationLocks.GetOrAdd(topicName, _ => new SemaphoreSlim(1, 1));
+            _logger.LogDebug("Publisher for topic {TopicName} not found, waiting on creation lock", topicName);
+            await lockObj.WaitAsync(cancellationToken).ConfigureAwait(false);
+            if (_batchPublisherCache.TryGetValue(topicName, out publisher))
+            {
+                _logger.LogDebug("Lock re-entrancy detected, returning existing publisher");
+                publisherTask.Add(publisher.PublishAsync(batchMessages, metadata, cancellationToken));
+                continue;
+            }
```

Comment:
I think there's a bug here, which is why that one test is failing.

If the batch contains messages that would be sent to different SNS topics based on the payload via the customisation delegate, then that is skipped and all the messages [end up in the same topic as the first message in the batch](https://github.com/justeattakeaway/JustSaying/actions/runs/7745312137/job/21120996624?pr=1335#step:6:369).

Reply:
Fixed by 11a93f01703404c8f2e3658c16e4e68e46e12164, though the test is still failing for some other reason.